src/app/clusters/color-control-server/color-control-server.cpp

```diff
@@ -47,711 +42,1080 @@
-static uint8_t addSaturation(uint8_t saturation1, uint8_t saturation2);
-static uint8_t subtractSaturation(uint8_t saturation1, uint8_t saturation2);
-static void initHueSat(EndpointId endpoint);
-static uint8_t readHue(EndpointId endpoint);
-static uint8_t readSaturation(EndpointId endpoint);
-static uint16_t addEnhancedHue(uint16_t hue1, uint16_t hue2);
-static uint16_t subtractEnhancedHue(uint16_t hue1, uint16_t hue2);
-static uint16_t readEnhancedHue(EndpointId endpoint);
-
-static bool moveHue(uint8_t moveMode, uint16_t rate, uint8_t optionsMask, uint8_t optionsOverride, bool isEnhanced);
-static bool moveToHue(uint16_t hue, uint8_t hueMoveMode, uint16_t transitionTime, uint8_t optionsMask, uint8_t optionsOverride,
-                      bool isEnhanced);
-static bool moveToHueAndSaturation(uint16_t hue, uint8_t saturation, uint16_t transitionTime, uint8_t optionsMask,
-                                   uint8_t optionsOverride, bool isEnhanced);
-static bool stepHue(uint8_t stepMode, uint16_t stepSize, uint16_t transitionTime, uint8_t optionsMask, uint8_t optionsOverride,
-                    bool isEnhanced);
+std::unordered_map<chip::EndpointId, ColorControlServer::ColorHueTransitionState> ColorControlServer::colorHueTransitionStates;
+std::unordered_map<chip::EndpointId, ColorControlServer::Color16uTransitionState> ColorControlServer::colorSatTransitionStates;
 #endif
 
 #ifdef EMBER_AF_PLUGIN_COLOR_CONTROL_SERVER_XY
-static uint16_t findNewColorValueFromStep(uint16_t oldValue, int16_t step);
-static uint16_t readColorX(EndpointId endpoint);
-static uint16_t readColorY(EndpointId endpoint);
-#endif
+std::unordered_map<chip::EndpointId, ColorControlServer::Color16uTransitionState> ColorControlServer::colorXtransitionStates;
+std::unordered_map<chip::EndpointId, ColorControlServer::Color16uTransitionState> ColorControlServer::colorYtransitionStates;
+#endif // EMBER_AF_PLUGIN_COLOR_CONTROL_SERVER_XY
+
+#ifdef EMBER_AF_PLUGIN_COLOR_CONTROL_SERVER_TEMP
+std::unordered_map<chip::EndpointId, ColorControlServer::Color16uTransitionState> ColorControlServer::colorTempTransitionStates;
+#endif // EMBER_AF_PLUGIN_COLOR_CONTROL_SERVER_TEMP
+
+std::unordered_map<chip::EndpointId, EmberEventControl> ColorControlServer::eventControls;
 
-static uint16_t computeTransitionTimeFromStateAndRate(Color16uTransitionState * p, uint16_t rate);
+/**********************************************************
+ * ColorControl Implementation
+ *********************************************************/
 
-// convenient token handling functions
-static uint8_t readColorMode(EndpointId endpoint)
+ColorControlServer & ColorControlServer::Instance()
 {
-    uint8_t colorMode;
-    emberAfReadServerAttribute(endpoint, ZCL_COLOR_CONTROL_CLUSTER_ID, ZCL_COLOR_CONTROL_COLOR_MODE_ATTRIBUTE_ID,
-                               reinterpret_cast<uint8_t *>(&colorMode), sizeof(uint8_t));
-    return colorMode;
+    return instance;
```

Comment:
The singleton pattern here for an instance of this class clashes with the need to have multiple instances of color control on each endpoint, resulting in the creation of global map instances above. This really does defeat the point of OO'fying logic in this file.

See my suggestions above for ways to fix this.

Reply:
Each endpoint doesn't need multiple instances of color control. The logic behind using a singleton is that the ColorControlSever class manages the transitionStates that are endpoint specific.

Even if the class wasn't a singleton, each endpoint would only need one ColorControlServer and one of each transitionState. The transitionStates would need to be persistent trought out the execution since they are needed in the calculations of the color variations.